logging: Add log_frontend support to v2

nordic-krch opened the pull request:
Added support for log_frontend in v2. Contrary to v1, in v2 frontend can coexist with backends. Frontend can be used to log directly to some high-speed hardware tracing module while logging to human readable backend is maintained. Logging to backends is suppressed when there is no backends registered or `CONFIG_LOG_FRONTEND_ONLY=y` is used. Contrary to backends, there can only be one frontend.

Added tests for frontend (v1 and v2) in `tests/subsys/logging/log_api` suite.

In order to implement frontend v2 following API must be implemented:
```
void log_frontend_init(void);
void log_frontend_panic(void);

/* Main logging function. */
void log_frontend_msg(const void *source, const struct log_msg2_desc desc,
                      uint8_t *package, const void *data);
```

It is based on #41971 thus DNM for now.

nordic-krch committed:
logging: Add frontend support to v2

Extended logging v2 to support frontend api. Contrary to v1,
it is possible to have frontend and backends in the system.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: logging: log_api: Extend with frontend feature

Extend test to check also frontend feature.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
doc: reference: logging: Extend frontend section

Update section about frontend feature.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@nashif this is suppose to fullfil your use case:
> is it possible to have 2 backends, for example RTT and UART where log messages over RTT are immediate and deferred/buffered over UART?

I figured out that it would be hard to do it with some backends processed in immediate mode and some in deferred. However, we have rarely used "frontend" feature that could be used here. It was not usable in v2. This PR brings it back with test coverage. In v2 it was fairly easy to make it work together with backends. In your case, you would need to implement dedicated frontend.

nordic-krch committed:
boards: posix: Cleanup backend config for native_posix and nrf52_bsim

Removing backend config from board Kconfig.defconfig and moving
it to native_posix backend configuration in logging. Without this
change define persisted even when logging was not using backends
and that impacted what is compiled in.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
soc: xtensa: sample_controller: Cleanup backend config

Removing backend config from board Kconfig.defconfig and moving
it to xtensa_sim backend configuration in logging. Without this
change define persisted even when logging was not using backends
and that impacted what is compiled in.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

dcpleung commented:
LGTM... Somehow I imagine someone complaining they could only do one frontend... :P

dcpleung commented on subsys/logging/Kconfig.mode:
```diff
@@ -36,14 +36,21 @@ config LOG_MODE_MINIMAL
 
-endchoice
+config LOG_FRONTEND_ONLY
+	bool "No backends"
```
Maybe `bool "No backends" if !LOG1`? Since v1 cannot have both frontend and backends enabled, removing the prompt prevents this from being disabled.

nordic-krch committed:
tests: logging: log_api: Fix test cases setup

Execute log_setup even when test is skipped. This is especially
vital when test infrastructure is using logging (e.g. LOG_PRINT
is enabled).

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@nashif @dcpleung only fixed documentation failure (missing ``kconfig:option:``).
